syntax = "proto3";

package buildle;

service Buildle {
    // Start a new game
    rpc Start (StartRequest) returns (StartResponse) {}
    // Attempt to guess a word
    rpc Guess (GuessRequest) returns (GuessResponse) {}
}

message StartRequest {
    bool hard = 1;
    string name = 2;
}

message StartResponse {
    int32 session = 1;
    int32 attempts = 2;
    int32 letters = 3;
}

message GuessRequest {
    int32 session = 1;
    string attempt = 2;
}

enum GuessResult {
    NO_MATCH = 0;  // Letter is not present in a word
    PARTIAL_MATCH = 1;  // Letter is present but misplaced
    MATCH = 2;  // Letter guessed correctly
}

message GuessResponseNormal {
    int32 attempts = 1;
    repeated GuessResult result = 2;
}

message GuessResponseLoss {
    string correct_word = 1;
    repeated GuessResult result = 2;
}

message GuessResponseWin {
}

message GuessResponse {
    oneof result {
        GuessResponseNormal keep_guessing = 1;
        GuessResponseLoss loss = 2;
        GuessResponseWin win = 3;
    }
}